E2115 choose mentor

AvlWx2014 opened the pull request:
* Add: new constant `Participant#DUTY_MENTOR` representing the duty title of mentor for the database. This will identify participants as mentors.
* Add: `MentorManagement` class, and class methods for the select mentor functionality.

AvlWx2014 committed:
Fix Markdown formatting in design_document.md

AvlWx2014 committed:
Add DUTY_MENTOR constant

AvlWx2014 committed:
Implement select_mentor and helper methods in new class MentorManagement

bahati2000 reviewed (commented):
I wonder what would be the trigger for this class. Because mentors will be assigned when an assignment without topics is created. Should the code to check when this is created be included in this class?  On second thought I don't think so. This might violate single responsibility principle.

AvlWx2014 committed:
Fix select mentor to return the selected mentor id or nil

Currently, it would return a tuple of [id, team_count] which is a detail callers don't need to know about.

AvlWx2014 committed:
Replace `first(1)` with `first` so unpacking works as expected

AvlWx2014 committed:
Merge branch 'beta' into e2115-choose-mentor

AvlWx2014 committed:
Add trigger to invitation.rb

AvlWx2014 committed:
Implement update_mentor_state

Commit:
wip mentor added by id

Nice peer programming session guys!

Co-Authored-By: AvlWx2014 <16199909+AvlWx2014@users.noreply.github.com>
Co-Authored-By: Hatsukikeika <43552712+Hatsukikeika@users.noreply.github.com>
Co-Authored-By: Bahati Wanza <bahati2000@users.noreply.github.com>

molinamelendezj commented:
@bahati2000 really wants these changes in beta lol getting blasted by approved emails 😆

Commit:
mentor views and instructor team adding trigger

another bites the dust!

Co-Authored-By: AvlWx2014 <16199909+AvlWx2014@users.noreply.github.com>
Co-Authored-By: Hatsukikeika <43552712+Hatsukikeika@users.noreply.github.com>

molinamelendezj committed:
missed null check

Co-Authored-By: AvlWx2014 <16199909+AvlWx2014@users.noreply.github.com>
Co-Authored-By: Hatsukikeika <43552712+Hatsukikeika@users.noreply.github.com>

molinamelendezj commented on app/models/mentor_management.rb:
```diff
@@ -0,0 +1,84 @@
+    User.where(id: mentor_user_id).first
+  end
+
+  def self.update_mentor_state(assignment_id, team_id)
```
considering rename to `assign_mentor()`?? add comments of what this method does.

molinamelendezj commented on app/models/mentor_management.rb:
```diff
@@ -0,0 +1,84 @@
+  # [assignment_id].
+  #
+  # @see participant.rb for the value of DUTY_MENTOR
+  def self.get_mentors_for_assignment(assignment_id)
```
consider rename

molinamelendezj commented on app/models/invitation.rb:
```diff
@@ -61,7 +61,8 @@ def self.accept_invite(team_id, inviter_user_id, invited_user_id, assignment_id)
     if can_add_member # The member was successfully added to the team (the team was not full)
       Invitation.update_users_topic_after_invite_accept(inviter_user_id, invited_user_id, assignment_id)
-
+      new_team_id = TeamsUser.team_id(assignment_id, inviter_user_id)
```
consider create issue?

AvlWx2014 committed:
Add auto_assign_mentor column to assignments

AvlWx2014 committed:
Start adding mentor checkbox to assignment form

AvlWx2014 committed:
rename file to match class name

AvlWx2014 committed:
rename the checkbox element

AvlWx2014 committed:
Match jQuery selector to checkbox id

molinamelendezj committed:
auto_assign_mentor flag logic

wip

Co-Authored-By: AvlWx2014 <16199909+AvlWx2014@users.noreply.github.com>
Co-Authored-By: Hatsukikeika <43552712+Hatsukikeika@users.noreply.github.com>

molinamelendezj committed:
added notify email, and feature toggle for mentor auto assign

Co-Authored-By: Bahati Wanza <bahati2000@users.noreply.github.com>
Co-Authored-By: AvlWx2014 <16199909+AvlWx2014@users.noreply.github.com>
Co-Authored-By: Hatsukikeika <43552712+Hatsukikeika@users.noreply.github.com>

bahati2000 committed:
added 5 tests for the MentorManagement methods

AvlWx2014 committed:
Add some more method-level comments

AvlWx2014 committed:
Passing user_a_mentor test

AvlWx2014 committed:
Replace 'context' blocks with 'it' blocks throughout

AvlWx2014 committed:
Passing select_mentor test

AvlWx2014 committed:
Remove notify_team_of_mentor_assignment test

AvlWx2014 committed:
Passing get_mentors_for_assignment test

AvlWx2014 committed:
Add 'it' block to last test so it actually gets picked up by RSpec

AvlWx2014 committed:
Should return empty array, rather than empty hash

AvlWx2014 committed:
Passing zip_mentors_with_team_count tests

AvlWx2014 committed:
Update test to reflect impl change

AvlWx2014 committed:
Update test to test for the right return result

AvlWx2014 commented on app/models/invitation.rb:
```diff
@@ -61,7 +61,8 @@ def self.accept_invite(team_id, inviter_user_id, invited_user_id, assignment_id)
       Invitation.update_users_topic_after_invite_accept(inviter_user_id, invited_user_id, assignment_id)
-
+      new_team_id = TeamsUser.team_id(assignment_id, inviter_user_id)
+      MentorManagement.update_mentor_state(assignment_id, new_team_id)
```
Add a comment for the presence of this code here.

AvlWx2014 commented on app/models/teams_user.rb:
```diff
@@ -6,7 +6,11 @@ class TeamsUser < ActiveRecord::Base
   def name(ip_address = nil)
-    self.user.name(ip_address)
+    name = self.user.name(ip_address)
+    if MentorManagement.user_a_mentor?(self.user)
```
Leave a comment for the reason this is here.

AvlWx2014 commented:
Merging so we can get this in to beta.
